test/utils/fakedns/fakedns.go

```diff
@@ -0,0 +1,129 @@
+// This is necessary because net.DefaultResolver is a global variable.
+var dnsHijackLock sync.Mutex
+
+// Server is a fake DNS server that can be used to override DNS resolution in tests.
+// This is particularly useful for testing components that rely on DNS without needing
+// to set up an actual DNS server. The server can be configured with a map of domain
+// names to IP addresses, and it will respond to DNS queries accordingly.
+// To use the fake DNS server, create an instance using NewServer, then call
+// Hijack to redirect DNS queries to the fake server.
+// This helper does not support running tests in parallel.
+type Server struct {
+	dnsServer forkednet.FakeDNSServer
+}
+
+// NewServer creates a new fake DNS server. The hosts map should contain
+// domain names as keys and their corresponding IP addresses as string values.
+// It returns an error if any of the IP addresses are invalid.
+func NewServer(hosts map[string]string) (*Server, error) {
+	parsedHosts := make(map[string]net.IP, len(hosts))
+	for domain, ipStr := range hosts {
+		ip := netutils.ParseIPSloppy(ipStr)
+		if ip == nil {
+			return nil, fmt.Errorf("invalid IP address for domain %s: %s", domain, ipStr)
+		}
+		// Ensure domain is clean without trailing dot for map key
+		parsedHosts[strings.TrimSuffix(domain, ".")] = ip
+	}
+	dnsServer := forkednet.FakeDNSServer{}
+	dnsServer.ResponseHandler = func(n, s string, q dnsmessage.Message, t time.Time) (dnsmessage.Message, error) {
+		resp := dnsmessage.Message{
+			Header: dnsmessage.Header{
+				ID:            q.Header.ID,
+				Response:      true,
+				Authoritative: true,
+			},
+			Questions: q.Questions,
+		}
+
+		if len(q.Questions) == 0 {
+			return resp, nil
+		}
+
+		question := q.Questions[0]
+		domain := strings.TrimSuffix(question.Name.String(), ".")
+
+		if ip, ok := parsedHosts[domain]; ok {
+			var resource dnsmessage.Resource
+			if ipv4 := ip.To4(); ipv4 != nil {
+				resource = dnsmessage.Resource{
+					Header: dnsmessage.ResourceHeader{Name: question.Name, Type: dnsmessage.TypeA, Class: dnsmessage.ClassINET, TTL: 300},
+					Body:   &dnsmessage.AResource{A: [4]byte(ipv4)},
+				}
+			} else { // Assume IPv6
+				resource = dnsmessage.Resource{
+					Header: dnsmessage.ResourceHeader{Name: question.Name, Type: dnsmessage.TypeAAAA, Class: dnsmessage.ClassINET, TTL: 300},
+					Body:   &dnsmessage.AAAAResource{AAAA: [16]byte(ip)},
+				}
+			}
+			resp.Answers = []dnsmessage.Resource{resource}
+		}
+
+		return resp, nil
+	}
+
+	return &Server{
+		dnsServer: dnsServer,
+	}, nil
+}
+
+// Hijack takes control of the net.DefaultResolver to redirect DNS queries
+// to the fake server. It uses t.Cleanup to automatically restore the original
+// resolver after the test.
+// This helper does not support running tests in parallel.
+//
+// Example:
+//
+//	dnsServer, _ := fakedns.NewServer(hosts)
+//	dnsServer.Hijack(t)
+func (s *Server) Hijack(t *testing.T) {
```

Comment:
extracted from the test as a helper because it can be very useful for integration tests, since we can only bind to localhost and that stops certain features to work, is not only environment proxy variables, localhost behaves different for Endpoints, Node connections and similar things,